contributors/design-proposals/encryption.md

```diff
@@ -0,0 +1,204 @@
+}
+```
+
+```go
+type KeyEncryptionKeySlot struct {
+    ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+    Slot int
+}
+```
+
+```go
+type KeyEncryptionKeyDatabase struct {
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+    Keys []string
+}
+```
+
+To enable encryption a user will issue a PUT to an endpoint such as `/rotate`. If this is the first time this API has ever been called the API server
+will generate a key (unecrypted DEK), encrypt it with the KEK in slot 1, and encrypt all secrets with the DEK. If this is the second + N time the API
+has been called the API server will encrypt the DEK with the KEK in slot N+1 and do a compare-and-swap on the DEK stored in etcd.
+If this API is called and a KEK doesn't exist in the KEK DB at slot N+1 a 400 is returned. `kubectl` will have support for calling the
+rotate API endpoint eventually. To account for failure scenarios during key rotation, the old and new DEK will be stored in etcd during the rotation.
+
+## Master Configuration
+In order to enable encryption, a user must first create a KEK DB file and tell the API server to use it with --key-encryption-key-db-path=/path/to/kekdb/file. The file will be a simple YAML file that lists all of the keys:
+
+```yaml
+kind: KeyEncryptionKeyDatabase
+version: v1
+keys:
+ - foo
+ - bar
+ - baz
+```
+
+The user will also need to specify the encryption provider and the resources to encrypt as follows:
+```yaml
+--key-encryption-key-db-path=/path/to/key-encryption-key/db
+--encryption-provider=default
+--encrypt-resource=v1/Secrets
+--encrypt-resource=v1/ConfigMap
+```
+
+As mentioned previously, the first pass at encryption will only support the default encryption provider.
+
+## Additional Considerations
+
+### Other encryption providers
+
+It should be easy to introduce new variations on the basic provider - such as those that require a password via STDIN on startup to decrypt the keys on disk, or wish to exchange a temporary token on disk for the actual keys from a third party security server.
+
+We expect that the provider implementation may therefore be layered or composed and that as much as possible providers should be separated behind interfaces.
+
+### Potential providers
+
+* A local HSM implementation that retrieves the keys from the secure enclave prior to reusing the AES-GCM implementation (initialization of keys only)
+* Exchanging a local temporary token for the actual decryption tokens from a networked secret vault
+* Decrypting the AES-256 keys from disk using asymmetric encryption combined with a user input password
+
+### Backwards Compatibility
+Once a user encrypts any resource in etcd, they are locked to that Kubernetes version and higher unless they choose to manually decrypt that resource in etcd. This will be discouraged. It will be highly recommended that users discern if their Kubernetes cluster is on a stable version before enabling encryption.
```

Comment:
I think this requires section on these two topics:

1. How rotation in storage happens
2. How encryption is enabled

Ideally 1 and 2 should follow the same pattern (adding a new key is a lot like turning on encryption at first).